xarray/core/computation.py

```diff
@@ -1371,16 +1371,14 @@ def _get_valid_values(da, other):
     da_b = da_b.map_blocks(_get_valid_values, args=[da_a])
 
     # 3. Detrend along the given dim
-    demeaned_da_a = da_a - da_a.mean(dim=dim)
-    demeaned_da_b = da_b - da_b.mean(dim=dim)
+    # https://github.com/pydata/xarray/issues/4804#issuecomment-760114285
+    demeaned_da_ab = (da_a * da_b) - (da_a.mean(dim=dim) * da_b.mean(dim=dim))
```

Comment:
shouldn't this be
```suggestion
    demeaned_da_ab = (da_a * da_b).mean(dim=dim) - da_a.mean(dim=dim) * da_b.mean(dim=dim)
```

Reply:
aaah yes sorry, one beer too many..

Reply:
@keewis actually no, it was right the first time I think! Because later on there's a `demeaned_da_ab.sum(dim=dim) ...`

Reply:
I think there was actually a typo/error in the original suggestion in #4804

Reply:
no, I think that was right: we're currently doing something like
```python
arr = (x - x.mean()) * (y - y.mean())
arr.sum(...) / valid_count
```
if we then use the suggestion from #4804 this becomes
```python
arr = (x * y).mean() - x.mean() * y.mean()
arr.sum(...) / valid_count
```

Reply:
beers all round, eh?!

Reply:
the only issue I still have is that
```python
a = np.random.random(10) + 10
b = np.random.random(10) + 15

x1 = (a - a.mean()) * (b - b.mean())
x2 = a * b - a.mean() * b.mean()

np.assert_allclose(x1, x2)
```
raises, so there's still something wrong

Reply:
apparently, this would need to become
```python
a * b - a * b.mean() - b * a.mean() + a.mean() * b.mean()
```

Reply:
Aaah yes, good point. I wonder if this still causes the same problem as was noted in [the original comment](https://github.com/pydata/xarray/issues/4804#issuecomment-760114285)? i.e. it "leads to Dask holding all chunks of `x` in memory"

Reply:
I guess that still has that issue.

Right, so actually that comment states that (note the additional `.mean()` in the left-hand side):
```python
((a - a.mean()) * (b - b.mean())).mean() = (a * b).mean() - a.mean() * b.mean()
```

Reply:
I think the issue might be that @willirath's suggestion was actually to combine the [_this whole operation_](https://github.com/pydata/xarray/blob/master/xarray/core/computation.py#L1374_L1383) into something like `x*y.mean() - x.mean()*y.mean()`. But it's not immediately obvious to me that that's what we want to do?

Reply:
Actually @keewis , what really matters is if `np.testing.assert_allclose(x1.sum(), x2.sum())` raises or not. We're not asserting that they're the same, but just that they give the same result when we `sum` over the appropriate dimension (which they do). This is why the tests are passing (so far)

Reply:
@AndrewWilliams3142 See here for a quick illustration of the issue: https://nbviewer.jupyter.org/gist/willirath/a555521c5b979509435fc247d69ea895